provider/db.go

```diff
@@ -0,0 +1,424 @@
+package provider
+
+import (
+	"database/sql"
+	"fmt"
+	"skybin/core"
+	"time"
+
+	_ "github.com/mattn/go-sqlite3" //sqlite library
+)
+
+func (p *Provider) SetupDB(path string) (*sql.DB, error) {
+	// Open DB
+	db, err := sql.Open("sqlite3", path)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open DB. error: %s", err)
+	}
+
+	// Create contracts table
+	stmt, err := db.Prepare(`CREATE TABLE IF NOT EXISTS contracts ( id INTEGER PRIMARY KEY, 
+		ContractId TEXT, 
+		RenterId TEXT, 
+		ProviderId TEXT, 
+		StorageSpace INTEGER, 
+		StartDate TEXT, 
+		EndDate TEXT, 
+		RenterSignature TEXT, 
+		ProviderSignature TEXT)`)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to prepare contracts table. error: %s", err)
+	}
+	_, err = stmt.Exec()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create contracts table. error: %s", err)
+	}
+
+	// Create blocks table
+	stmt, err = db.Prepare(`CREATE TABLE IF NOT EXISTS blocks ( id INTEGER PRIMARY KEY, 
+		RenterId TEXT, 
+		BlockId TEXT, 
+		Size INTEGER)`)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to prepare blocks table. error: %s", err)
+	}
+	_, err = stmt.Exec()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create blocks table. error: %s", err)
+	}
+
+	// Create activity table
+	stmt, err = db.Prepare(`CREATE TABLE IF NOT EXISTS activity ( id INTEGER PRIMARY KEY, 
```

Comment:
The activity handling is clever but, I think, a little wonky. You also lose all of the counts for the previous period when you roll over (correct me if I'm wrong?).

Here's what I recommend you do. I think this is more similar to what you were previously doing:

Keep a top-level struct for the activity feed. In this, keep counters for everything that happened in the last 24 hours on a per-hour basis and in the last 7 days on a per-day basis.

```
type ActivityFeed struct {

	// Counters for the last 24 hours. Typically, len(hourlyCounts) == 7.
	hourlyCounts []activityCounts

	// Daily counts for the last week. Typically, len(dailyCounts) == 7.
	dailyCounts  []activityCounts
}

type activityCounts struct {

	// The timestmap for this set of counters, truncated down to the
	// nearest hour (for hourlyCounters) or day (for daily counters)
	TimeStamp           time.Time
	BlockUploads        int64
	BlockDownloads      int64
	BlockDeletions      int64
	StorageReservations int64
	BytesUploaded       int64
	BytesDownloaded     int64
}
```


Add a method on the feed which updates the counters when a new entry is ready. Take the current timestamp as a parameter to make it easier to test. The method will look roughly like this (there are some edge cases I don't handle correctly, but this is the general gist). 

```
func (feed *ActivityFeed) AddOp(t time.Time, op string, byteAmt int64) {

	// Lop off the old hourly counters from more than 24 hours ago
	idx := 0
	hour := t.Truncate(time.Hour)
	for idx < len(feed.hourlyCounts) && hour.Sub(feed.hourlyCounts[idx].TimeStamp) > 24 * time.Hour {
		idx++
	}
	feed.hourlyCounts = feed.hourlyCounts[idx:]

	// Lop off old daily counters from more than 7 days ago
	idx = 0
	day := t.Truncate(24 * time.Hour)
	for idx < len(feed.dailyCounts) && day.Sub(feed.dailyCounts[idx].TimeStamp) > 7 * 24 * time.Hour {
		idx++
	}
	feed.dailyCounts = feed.dailyCounts[idx:]

	// Append new hourly counters for all the hours between the last counter and now.
	// You'll need to watch for the edge case where len(feed.hourlyCounts) == 0!!!
	for len(feed.hourlyCounts) < 24 {
		lastTimestamp := feed.hourlyCounts[len(feed.hourlyCounts) - 1].TimeStamp
		nextTimestamp := lastTimestamp.Add(time.Hour)
		feed.hourlyCounts = append(feed.hourlyCounts, activityCounts{TimeStamp: nextTimestamp})
	}

	// Ditto here for the daily counters
	for len(feed.dailyCounts) < 7 {

	}

        // Now actually update the counts for the most recent period.
	hourlyCounts := &feed.hourlyCounts[len(feed.hourlyCounts) - 1]
	dailyCounts := &feed.dailyCounts[len(feed.dailyCounts) - 1]
	countsToIncrement := []*activityCounts{hourlyCounts, dailyCounts}

	switch op {
	case "BlockUpload":
		for _, counts := range countsToIncrement {
			counts.BlockUploads++
			counts.BytesUploaded += byteAmt
		}
	//	... And so on and so forth for the different types ...
	}
}
```

Then, add a method to generate the summary expected by the GET /stats endpoint. I'll leave the details to you.

```
func (feed *ActivityFeed) GenerateSummary() *ActivitySummary
```

Then add unit tests for the activity feed.

Finally, DON'T save the activity feed in SQLite. Just put it in a plain old JSON file. Update the code in the cmd/provider.go to handle OS signals. When you get a signal to shutdown, stop the server and save the snapshot file. I've done this a few times before and it looks roughly like this:

```
//  ... inside the provider daemon command
go func() {
  signal := <- os.Signal
  provider.SaveSnapshot()   

  // You'll have to look into this part. Worst case, just call provider.SaveSnapshot() then os.Exit() and don't worry
  // about cleanly shutting down the server.
  server.Shutdown()
  os.Exit(0)
}()
```

Reply:
Alternatively, I think leveraging SQLite could make things much easier and more robust. Rather than maintaining counters and worrying about locks and such, I would just log all actions to the database, then use database operations (which should be pretty fast) to get the stats we want.

For example, you could use a query like the one mentioned [here](https://stackoverflow.com/questions/10643379/how-do-i-query-for-all-dates-greater-than-a-certain-date-in-sql-server) to get all of the entries newer than a certain day (say, last week), then just use COUNT to get the specific number of that type of action.

I prefer that approach because we don't have to do a lot of complicated logic, let the database do the heavy lifting, and collect more data, which gives us the option to pull more stats out of existing data later if we choose.

Reply:
It cycles in the same manner that the old one does, the only difference is that it is stored in SQLite as opposed to a snapshot.

It might be unclear in some of the code (I added some comments) but 'day' intervals correspond to 24 1-hour intervals and they cycle out of scope if they are more than 24-hours old.

I would like to switch to the method Kincaid suggested however I think it would struggle on the UploadedBytes and DownloadedBytes as they would have to be summed as opposed to using the COUNT method.